util/election.go

```diff
@@ -0,0 +1,112 @@
+package util
+
+import (
+	"fmt"
+	"github.com/timescale/prometheus-postgresql-adapter/log"
+	"net/http"
+	"sync"
+)
+
+// Election defines an interface for adapter leader election.
+// If you are running Prometheus in HA mode where each Prometheus instance sends data to corresponding adapter you probably
+// want to allow writes into the database from only one adapter at the time. We need to elect a leader who can write to
+// the database. If leader goes down, another leader is elected. Look at `lock.go` for an implementation based on PostgreSQL
+// advisory locks. Should be easy to plug in different leader election implementations.
+type Election interface {
+	Id() string
+	BecomeLeader() (bool, error)
+	IsLeader() (bool, error)
+	Resign() error
+}
+
+// RestElection is a REST interface allowing to plug in any external leader election mechanism.
+// Remote service can use REST endpoints to manage leader election thus block or allow writes.
+type RestElection struct {
+	leader bool
+	mutex  sync.RWMutex
+}
+
+func NewRestElection() *RestElection {
+	r := &RestElection{}
+	http.Handle("/admin/election/leader", r.handleLeader())
+	http.Handle("/admin/election/resign", r.handleResign())
+	return r
+}
+
+func (r *RestElection) handleLeader() http.HandlerFunc {
+	return func(response http.ResponseWriter, request *http.Request) {
+		switch request.Method {
+		case http.MethodGet:
+			leader, err := r.IsLeader()
+			if err != nil {
+				log.Error("msg", "Failed on leader check", "err", err)
+				http.Error(response, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			fmt.Fprintf(response, "%v", leader)
+		case http.MethodPost:
+			leader, err := r.BecomeLeader()
+			if err != nil {
+				log.Error("msg", "Failed to become a leader", "err", err)
+				http.Error(response, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			fmt.Fprintf(response, "%v", leader)
+		default:
+			log.Error("msg", "Request method not supported")
+			http.Error(response, "Request method not supported", http.StatusInternalServerError)
+		}
+	}
+}
+
+func (r *RestElection) handleResign() http.HandlerFunc {
+	return func(response http.ResponseWriter, request *http.Request) {
+		if request.Method == http.MethodPost {
+			err := r.Resign()
+			if err != nil {
+				log.Error("err", err)
+				http.Error(response, "Failed to resign", http.StatusInternalServerError)
+				return
+			}
+			response.WriteHeader(http.StatusOK)
+		} else {
+			log.Error("msg", "Request method not supported")
+			http.Error(response, "Request method not supported", http.StatusInternalServerError)
+		}
+
+	}
+}
+
+func (r *RestElection) Id() string {
+	return ""
+}
+
+func (r *RestElection) BecomeLeader() (bool, error) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	if r.leader {
+		log.Debug("msg", "Instance is already a leader")
+		return r.leader, nil
+	}
+	r.leader = true
+	log.Info("msg", "Instance became a leader")
```

Comment:
FWIW, I would wrap the implementations of this interface in a higher-level object that does the logging for any implementation.  Alternatively, just log where called. Then the logging is consistent across implementations.

For instance, you could have an `Elector` type that calls into an `Election` implementation and does the logging depending on return values from the specific implementation. So, if `BecomeLeader()` returns true, then the `Elector` logs `"Instance became a leader"`